### PR TITLE
loading screen for terria

### PR DESCRIPTION
**Pull request by chloeleichen**
There are three source of delay:

1. downloading the massive `terriamap.js`
2. react getting data to initialise dom
3. cesium/leaflet loading the map

This PR address the first issue, the rest two need to be looked at separately.

cc @philipgrimmett

### CODE REVIEW

File: buildprocess/webpack.config.js [hunk 1]
```diff
@@ -8,10 +8,13 @@ var path = require('path');
 
 module.exports = function(devMode, hot) {
     var config = {
-        entry: './index.js',
+        entry: {
+          index: './index.js',
```
[the remark this answers is not in the document]
Comment (chloeleichen) [reply buildprocess/webpack.config.js]: you are right :D

**Comment by chloeleichen** [comment]
@kring  `require.ensure` uses promise behind the scene. I guess we need to polyfill it in entry.js?

File: buildprocess/webpack.config.js [hunk 2]
```diff
@@ -8,10 +8,12 @@ var path = require('path');
...
 module.exports = function(devMode, hot) {
     var config = {
-        entry: './index.js',
+        entry: {
+          entry: './entry.js'
+         },
         output: {
             path: path.resolve(__dirname, '..', 'wwwroot', 'build'),
-            filename: 'TerriaMap.js',
+            filename: '[name].bundle.js',
```
Comment (kring) [inline buildprocess/webpack.config.js]: Can we revert this change? And then we can revert the index.html change too, which will be good because it will potentially avoid conflicts for our users in that file.

File: buildprocess/webpack.config.js [hunk 3]
```diff
@@ -8,10 +8,12 @@ var path = require('path');
 
 module.exports = function(devMode, hot) {
     var config = {
-        entry: './index.js',
+        entry: {
+          entry: './entry.js'
```
Comment (kring) [inline buildprocess/webpack.config.js]: Maybe switch back to the simpler form here.

**Comment by kring** [comment]
> require.ensure uses promise behind the scene. I guess we need to polyfill it in entry.js?

**Comment by kring** [comment]
I think we can just put this code into entry.js:
https://github.com/TerriaJS/terriajs/blob/master/lib/Models/Terria.js#L85

That will mean that entry.js will need to require-in `Promise.js` from TerriaJS, which will make the built entry.js a little bigger, but only a little.

**Comment by chloeleichen** [comment]
@kring ready for another look!

**Comment by kring** [comment]
This is so awesome!

I tweaked a few things:
- Used a promise instead of a callback in the loader.
- Embedded the GIF and CSS in the initial javascript to reduce the number of requests.
- Made index.js return the promise returned by `terria.start`, and made the loader wait on it before hiding the loader. This eliminates the second source of delay in your list at the top of this PR (well, it doesn't eliminate it, but it keeps showing the loader until it's done rather than flashing white during this process).

I think it's still loading the entire Terria CSS (35.9KB gzipped, 353KB uncompressed) and Google Fons (pretty small) before showing the loader, so that's something to improve, but I can't think of an easy way to do it. This is still a hundred times better than what we had before.
